### PR TITLE
Feature refactor materials

### PR DESCRIPTION
**Pull request by Nielsbishere**
Materials are now refactored and don't store duplicated data / don't have duplicated functions. This fixes textureless materials for RT.

### CODE REVIEW

File: src/material_pool.cpp [hunk 1]
```diff
@@ -12,218 +12,85 @@ namespace wr
...
 		}
 		else
 		{
-			if (!m_texture_pool)
-			{
-				m_texture_pool = ao.m_pool;
-			}
-			m_ao = ao;
-			m_material_data.m_material_flags.m_has_ao_texture = true;
+			m_textures[size_t(type) % size_t(MaterialTextureType::COUNT)] = handle;
```
Comment (LaisoEmilio) [inline src/material_pool.cpp]: not a fan of the C-style casting everywhere honestly

File: src/material_pool.cpp [hunk 2]
```diff
@@ -12,218 +12,85 @@ namespace wr
...
+	void Material::SetConstant(MaterialConstantType type, float val) {
+		float arr[1] = { val };
+		SetConstant(type, arr);
 	}
 
-	void Material::SetConstantRoughness(float roughness)
-	{
-		m_material_data.m_metallic_roughness = DirectX::XMVectorSetW(m_material_data.m_metallic_roughness, roughness);
-		m_material_data.m_material_flags.m_has_roughness_constant = true;
+	void Material::GetConstant(MaterialConstantType type, float &val) {
```
Comment (LaisoEmilio) [inline src/material_pool.cpp]: we already talked in person about this, it's a bit meh

File: src/material_pool.hpp [hunk 3]
```diff
@@ -11,157 +11,126 @@
...
-		TextureHandle GetNormal() { return m_normal; }
-		void SetNormal(TextureHandle normal);
+		TextureHandle GetTexture(MaterialTextureType type);
+		void SetTexture(MaterialTextureType type, TextureHandle handle);
+		void ClearTexture(MaterialTextureType type);
+		bool HasTexture(MaterialTextureType type);
 
-		void UseNormalTexture(bool use_normal);
+		template<uint16_t s>
+		void GetConstant(MaterialConstantType type, float(&val)[s]);
```
Comment (LaisoEmilio) [inline src/material_pool.hpp]: aaaaaa. You know what we don't like about this

File: demo/scene_emibl.hpp [hunk 4]
```diff
@@ -95,107 +95,107 @@ namespace emibl_scene
...
-				rusty_metal_material = material_pool->Create();
+				rusty_metal_material = material_pool->Create(texture_pool.get());
 
 				wr::Material* rusty_metal_internal = material_pool->GetMaterial(rusty_metal_material);
 
-				rusty_metal_internal->SetAlbedo(metal_splotchy_albedo);
-				rusty_metal_internal->SetNormal(metal_splotchy_normal);
-				rusty_metal_internal->SetRoughness(metal_splotchy_roughness);
-				rusty_metal_internal->SetMetallic(metal_splotchy_metallic);
+				rusty_metal_internal->SetTexture(wr::MaterialTextureType::ALBEDO, metal_splotchy_albedo);
```
Comment (DBouma) [inline demo/scene_emibl.hpp]: Why remove the explicit calls to setting the type of material? "SetAlbedo" is much more clear than calling SetTexture(MaterialTextureType::ALBEDO. Consider using more explicit naming

File: src/d3d12/d3d12_renderer.cpp [hunk 5]
```diff
@@ -1010,7 +1010,7 @@ namespace wr
 			texture_pool = static_cast<D3D12TexturePool*>(m_texture_pools[0].get());
 		}
 
-		auto albedo_handle = material_internal->GetAlbedo();
+		auto albedo_handle = material_internal->GetTexture(MaterialTextureType::ALBEDO);
```
Comment (DBouma) [inline src/d3d12/d3d12_renderer.cpp]: Consider explicit naming as stated before.

File: src/material_pool.cpp (hunk 1 above)
Comment (DBouma) [reply src/material_pool.cpp]: Try to be more explicit in your casts as well, increases readability in general.

File: src/material_pool.hpp [hunk 6]
```diff
@@ -8,160 +8,163 @@
...
+		Material(TexturePool* pool);
+		Material(TexturePool* pool,
+				 TextureHandle albedo, 
 				 TextureHandle normal, 
 				 TextureHandle roughness,
 				 TextureHandle metallic,
 				 TextureHandle ao, 
-				 bool alpha_masked = false, 
-				 bool double_sided = false);
+				 float alpha_constant = false,
```
Comment (LaisoEmilio) [inline src/material_pool.hpp]: talked in person about this

File: src/material_pool.hpp [hunk 7]
```diff
@@ -8,160 +8,163 @@
...
 
-		void SetConstantMetallic(DirectX::XMFLOAT3 metallic);
+				float m_color[3];
+				float m_metallic;
 
-		bool UsesConstantMetallic() { return m_material_data.m_material_flags.m_use_metallic_constant; };
-		void SetUseConstantMetallic(bool use_constant_metallic) { m_material_data.m_material_flags.m_use_metallic_constant = use_constant_metallic; };
+				float m_roughness;
+				float m_is_double_sided;
+				float m_use_alpha_constant;
```
Comment (LaisoEmilio) [inline src/material_pool.hpp]: same as before, this is alpha masked not constant, if you want to support both then we need two flags

File: demo/scene_emibl.hpp (hunk 4 above)
Comment (Nielsbishere) [reply demo/scene_emibl.hpp]: It leaves a lot of duplicated code, makes it harder to add an additional texture type. Now you just add to the enum.

File: src/d3d12/d3d12_renderer.cpp (hunk 5 above)
Comment (Nielsbishere) [reply src/d3d12/d3d12_renderer.cpp]: See previous comment

File: src/material_pool.hpp [hunk 8]
```diff
@@ -8,160 +8,163 @@
...
+		METALLIC = (3 << 16) | 0,
+		ROUGHNESS = (4 << 16) | 0,
+		IS_DOUBLE_SIDED = (5 << 16) | 0,
+		USE_ALPHA_CONSTANT = (6 << 16) | 0,
+		MAX_OFFSET = 7
+	};
+
+
+	template<MaterialConstantType type, bool b = uint32_t(type) & 0x1>
+	struct MaterialConstantHelper
```
Comment (Nielsbishere) [inline src/material_pool.hpp]: Viktor: Use std::enable_if instead.

**Pull request merged**